alacritty_terminal/src/tty/windows/conpty.rs

```diff
@@ -20,9 +23,63 @@ use crate::event::{OnResize, WindowSize};
 use crate::tty::windows::child::ChildExitWatcher;
 use crate::tty::windows::{cmdline, win32_string, Pty};
 
+/// Load the pseudoconsole API from conpty.dll if possible, otherwise use the
+/// standard Windows API. The conpty.dll from the Windows Terminal project
+/// supports loading OpenConsole.exe, which offers many improvements and
+/// bugfixes compared to the standard conpty that ships with Windows.
+///
+/// In particular, the new conpty.dll and OpenConsole.exe enable mouse
+/// reporting, which not yet supported by the Windows API on Windows 10 as of
+/// June 2023.
+///
+/// The conpty.dll and OpenConsole.exe files will be searched in PATH and in
+/// the directory where Alacritty executable is located.
+struct ConptyApi {
+    create: unsafe extern "system" fn(COORD, HANDLE, HANDLE, u32, *mut HPCON) -> HRESULT,
+    resize: unsafe extern "system" fn(HPCON, COORD) -> HRESULT,
+    close: unsafe extern "system" fn(HPCON),
+}
+
+impl ConptyApi {
+    /// Try loading ConptyApi from conpty.dll library.
+    fn load_conpty() -> Option<Self> {
+        // Unsafe because of Windows API calls.
+        unsafe {
+            let hmodule = LoadLibraryA(s!("conpty.dll"));
+            if hmodule == 0 {
+                return None;
+            }
+            let create_fn = GetProcAddress(hmodule, s!("CreatePseudoConsole"))?;
+            let resize_fn = GetProcAddress(hmodule, s!("ResizePseudoConsole"))?;
+            let close_fn = GetProcAddress(hmodule, s!("ClosePseudoConsole"))?;
+
+            Some(Self {
+                create: mem::transmute(create_fn),
+                resize: mem::transmute(resize_fn),
+                close: mem::transmute(close_fn),
```

Comment:
What are the actual types here? You probably shouldn't be using transmute.

Reply:
I would be happy to use something less intrusive, but I don't know if there is a way. Quoting the error message without the `transmute`:

```
error[E0308]: mismatched types
  --> alacritty_terminal\src\tty\windows\conpty.rs:54:25
   |
54 |                 create: create_fn,
   |                         ^^^^^^^^^ incorrect number of function parameters
   |
   = note: expected fn pointer `unsafe extern "system" fn(COORD, isize, isize, u32, *mut isize) -> i32`
              found fn pointer `unsafe extern "system" fn() -> isize`
```

The other three functions are also reported to have signature `unsafe extern "system" fn() -> isize`

Reply:
Given that those are pointers, try `as _`. `std::mem::transmute` shouldn't be used on pointers generally, because you have alternative like I've mentioned, and it's error prone if you change pointer to something else.

Reply:
It doesn't work. I defined the needed types and made a function `build()` to allow the `as _` syntax, but the cast is still not allowed:
```
error[E0605]: non-primitive cast: `unsafe extern "system" fn() -> isize` as `unsafe extern "system" fn(COORD, isize, isize, u32, *mut isize) -> i32`
  --> alacritty_terminal\src\tty\windows\conpty.rs:61:30
   |
61 |             Some(Self::build(create_fn as _, resize_fn as _, close_fn as _))
   |                              ^^^^^^^^^^^^^^ invalid cast
```
Please see https://stackoverflow.com/questions/46134477/how-can-i-call-a-raw-address-from-rust

Reply:
Hm, yes.

Could we do something like that then:

```diff
diff --git a/alacritty_terminal/src/tty/windows/conpty.rs b/alacritty_terminal/src/tty/windows/conpty.rs
index 8770aa6b..e780a46c 100644
--- a/alacritty_terminal/src/tty/windows/conpty.rs
+++ b/alacritty_terminal/src/tty/windows/conpty.rs
@@ -23,6 +23,31 @@ use crate::event::{OnResize, WindowSize};
 use crate::tty::windows::child::ChildExitWatcher;
 use crate::tty::windows::{cmdline, win32_string, Pty};
 
+type LoadedFn = unsafe extern "system" fn() -> isize;
+type CreatePseudoConsoleFn =
+    unsafe extern "system" fn(COORD, HANDLE, HANDLE, u32, *mut HPCON) -> HRESULT;
+type ResizePseudoConsoleFn = unsafe extern "system" fn(HPCON, COORD) -> HRESULT;
+type ClosePseudoConsoleFn = unsafe extern "system" fn(HPCON);
+
+/// RAII Pseudoconsole.
+pub struct Conpty {
+    pub handle: HPCON,
+    api: ConptyApi,
+}
+
+impl Drop for Conpty {
+    fn drop(&mut self) {
+        // XXX: This will block until the conout pipe is drained. Will cause a deadlock if the
+        // conout pipe has already been dropped by this point.
+        //
+        // See PR #3084 and https://docs.microsoft.com/en-us/windows/console/closepseudoconsole.
+        unsafe { (self.api.close)(self.handle) }
+    }
+}
+
+// The ConPTY handle can be sent between threads.
+unsafe impl Send for Conpty {}
+
 /// Load the pseudoconsole API from conpty.dll if possible, otherwise use the
 /// standard Windows API.
 ///
@@ -33,12 +58,24 @@ use crate::tty::windows::{cmdline, win32_string, Pty};
 /// The conpty.dll and OpenConsole.exe files will be searched in PATH and in
 /// the directory where Alacritty's executable is located.
 struct ConptyApi {
-    create: unsafe extern "system" fn(COORD, HANDLE, HANDLE, u32, *mut HPCON) -> HRESULT,
-    resize: unsafe extern "system" fn(HPCON, COORD) -> HRESULT,
-    close: unsafe extern "system" fn(HPCON),
+    create: CreatePseudoConsoleFn,
+    resize: ResizePseudoConsoleFn,
+    close: ClosePseudoConsoleFn,
 }
 
 impl ConptyApi {
+    /// Return the conpty API.
+    fn new() -> Self {
+        Self::load_conpty().unwrap_or({
+            // Cannot load conpty.dll - use the standard Windows API.
+            Self {
+                create: CreatePseudoConsole,
+                resize: ResizePseudoConsole,
+                close: ClosePseudoConsole,
+            }
+        })
+    }
+
     /// Try loading ConptyApi from conpty.dll library.
     fn load_conpty() -> Option<Self> {
         unsafe {
@@ -51,45 +88,14 @@ impl ConptyApi {
             let close_fn = GetProcAddress(hmodule, s!("ClosePseudoConsole"))?;
 
             Some(Self {
-                create: mem::transmute(create_fn),
-                resize: mem::transmute(resize_fn),
-                close: mem::transmute(close_fn),
+                create: mem::transmute::<LoadedFn, CreatePseudoConsoleFn>(create_fn),
+                resize: mem::transmute::<LoadedFn, ResizePseudoConsoleFn>(resize_fn),
+                close: mem::transmute::<LoadedFn, ClosePseudoConsoleFn>(close_fn),
             })
         }
     }
-
-    /// Return the conpty API.
-    fn new() -> Self {
-        Self::load_conpty().unwrap_or({
-            // Cannot load conpty.dll - use the standard Windows API.
-            Self {
-                create: CreatePseudoConsole,
-                resize: ResizePseudoConsole,
-                close: ClosePseudoConsole,
-            }
-        })
-    }
 }
 
-/// RAII Pseudoconsole.
-pub struct Conpty {
-    pub handle: HPCON,
-    api: ConptyApi,
-}
-
-impl Drop for Conpty {
-    fn drop(&mut self) {
-        // XXX: This will block until the conout pipe is drained. Will cause a deadlock if the
-        // conout pipe has already been dropped by this point.
-        //
-        // See PR #3084 and https://docs.microsoft.com/en-us/windows/console/closepseudoconsole.
-        unsafe { (self.api.close)(self.handle) }
-    }
-}
-
-// The ConPTY handle can be sent between threads.
-unsafe impl Send for Conpty {}
-
 pub fn new(config: &PtyConfig, window_size: WindowSize) -> Option<Pty> {
     let api = ConptyApi::new();
     let mut pty_handle: HPCON = 0;

```